pkg/scheduler/reconcile_on_synctarget.go

```diff
@@ -19,20 +19,220 @@ package scheduler
 import (
 	"context"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/klog/v2"
 
 	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
 )
 
-func (c *controller) reconcileOnSyncTarget(ctx context.Context, key string) error {
+func (c *controller) reconcileOnSyncTarget(ctx context.Context, stKey string) error {
 	logger := klog.FromContext(ctx)
-	ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	stws, _, stName, err := kcpcache.SplitMetaClusterNamespaceKey(stKey)
 	if err != nil {
-		logger.Error(err, "invalid key")
+		logger.Error(err, "invalid SyncTarget key")
 		return err
 	}
-	logger = logger.WithValues("workspace", ws, "syncTarget", name)
+	logger = logger.WithValues("workspace", stws, "syncTarget", stName)
 	logger.V(2).Info("reconciling")
 
+	/*
+		On synctarget change:
+		1) find all its loc(s), and update store
+
+		2a) find all its ep(s), and update store
+		2b) for each of its obsolete ep, remove all sp(s) that has st
+		2c) for each of its ongoing ep, update sp(s)
+		2d) for each of its new ep, add sp(s)
+
+		Need data structure:
+		- map from a synctarget to its locations, say 'locsBySelectedSt'
+		- map from a synctarget to its eps, say 'epsByUsedSt'
+	*/
+
+	st, _ := c.synctargetLister.Cluster(stws).Get(stName) // TODO: handle err
+
+	// 1)
+	locsColocating, _ := c.locationLister.Cluster(stws).List(labels.NewSelector()) // TOOD: handle err
+	locsSelectingSt, _ := filterLocsBySt(locsColocating, st)                       // TODO: handle err
+
+	locKeys := extractLocKeys(locsSelectingSt)
+	// locsSelectedSt := store.locsBySelectedSt[stKey]
+	store.locsBySelectedSt[stKey] = locKeys
+
+	// 2a)
+	epsUsingSt := []string{}
+	for _, loc := range locsSelectingSt {
+		locKey := kcpcache.ToClusterAwareKey(string(logicalcluster.From(loc)), loc.GetNamespace(), loc.GetName())
```

Comment:
Why not simply use `kcpcache.MetaClusterNamespaceKeyFunc(loc)` here?  BTW, you are guaranteed that it will never return an error when given a `Location`.

Reply:
I saw `MetaClusterNamespaceKeyFunc` calls `ToClusterAwareKey` so tried the latter out. Changed it back for simplicity.

May I know why an error is never returned here?